peps/pep-0735.rst

```diff
@@ -0,0 +1,476 @@
+
+* A string, which is a valid PEP 735 Dependency Specifier. e.g., ``numpy>1``
+* An object, which has exactly one key, ``spec``, which is a valid PEP 735
+  Dependency Spec. e.g., ``{spec = "numpy>1"}``
+
+Any additional keys in a requirement specifier object are reserved for future
+use.
+
+The following is an example of a ``pyproject.toml`` section which uses this to
+define four dependency groups: ``test``, ``docs``, ``typing``, and
+``typing-test``:
+
+.. code:: toml
+
+    [requirements.packages]
+    test = ["pytest", "coverage", "."]
+    docs = ["sphinx", "sphinx-rtd-theme"]
+    typing = ["mypy", "types-requests", ".[types]"]
+    typing-test = ["[typing,test]", "useful-types"]
+
+Note how ``test`` and ``typing`` are able to refer to the current package
+while ``docs`` does not. This reflects the ability of Dependency Groups to be used
+in the same manner as extras, adding to dependencies, or completely
+independently. ``typing-test`` is defined as a union of two existing groups,
+plus an additional package. ``typing`` includes an extra, ``types``.
+
+Implementation Requirements
+---------------------------
+
+Build backends MUST NOT include dependency group data in built distributions.
+
+Tools which support Dependency Groups MUST support both string and object
+representations of requirements.
+
+When unrecognized keys are encountered in requirement specifiers or the
+``requirements`` table, tools MUST NOT fail. They MAY emit warnings.
+
+When installing the current package (``.``) from a dependency group, tools
+SHOULD prefer editable installs over non-editable installs. They MAY provide
+users with options to configure this behavior.
+
+Use of Dependency Groups
+------------------------
+
+Tools which support Dependency Groups are expected to provide new options and
+interfaces to allow users to install from Dependency Groups. Implementations
+may wish to treat them similarly to their current treatments of
+``requirements.txt`` files, or more similarly to extras.
+
+No syntax is defined for expressing the dependency group of a package, for two
+reasons:
+
+* it would not be valid to refer to the dependency groups of a third-party
+  package from pypi (because the data is defined to be unpublished)
+
+* there is not guaranteed to be a current package for dependency groups -- part
+  of their purpose is to support non-package projects
+
+For example, a possible ``pip`` interface for installing dependency groups
+would be:
+
+.. code:: shell
+
+    pip install --dependency-groups=test,typing
+
+Note that this is only an example. This PEP does not declare any requirements
+for how tools support the installation of Dependency Groups.
+
+Reference Implementation
+========================
+
+TODO! STUB!
+
+Planned reference implementation:
+A very simple "environment manager" which can be used to build virtualenvs from
+specified Dependency Groups. It will therefore need to fully support parsing of
+Dependency Groups.
```

Comment:
```suggestion
There is currently no reference implementation/consumer of this specification.
```

Reply:
Okay, I wasn't sure if I should describe what I plan to add or not.
Perhaps a reference implementation isn't important in this case? I had expected it would be valuable to include one.